### ISSUE TITLE
test_db data in combined_graph_data

### ISSUE THREAD

**Issue by tonytan4ever** [NVIDIA/DIGITS#760]
According to:
    https://github.com/NVIDIA/DIGITS/blob/master/digits/model/tasks/train.py#L76~L77

   And https://github.com/NVIDIA/DIGITS/blob/master/digits/model/tasks/train.py#L457~L529

trained output (should from trained db) and validation output(from validation db) are being graphed. If I want to also graph test output, how can I achieve that ?

Thanks!.

**Comment by tonytan4ever** [comment]
Also, as we are using caffe framework,  

I am assuming this file:
https://github.com/NVIDIA/DIGITS/blob/master/digits/model/tasks/caffe_train.py needs to be add more layers to include test DB data, right ?

**Comment by lukeyeager** [comment]
Graphing test data during training doesn't really make sense and is not something we'll be adding to DIGITS. You could have multiple validation sets, but that's not a common use case.

If we were going to do anything with the test set in DIGITS, it would be along the lines of https://github.com/NVIDIA/DIGITS/pull/638, where you can test a model after training is already completed.

**Comment by iwalkdaline** [comment]
I'm not sure why a real-time test wouldn't make sense.  Can you elaborate because the idea is to get a feeling for how the model does on a validation set does is taken from training and a test set of the same data taken from elsewhere?  Are you saying that doesn't make sense or are you saying that would just be constituted as a second validation set?  In practice, we have used a validation set that is more appropriately as test set.  The issue with this is that line may say accuracy and that's what we're interested in but it provides no insights on what's going wrong, i.e. the real world test data could be bad or just not aligned to training data very well.  I would agree that running a test in the traditional ML sense seems most logical but waiting days to get this feedback isn't idea.  We can implement this ourselves but we're looking for pointers on the best integration approach.

**Comment by lukeyeager** [comment]
Yes, I think we're just disagreeing on the nomenclature. I tend to think of any data used to verify a model while the model is still being trained as "validation" data, while I tend to think of data used to verify a trained model as "test" data. But it doesn't matter much what you call it.

Are you asking for DIGITS to support testing multiple "validation" (in the sense that I've defined it) sets?

Caffe can certainly do it, and it wouldn't be hard to add for Torch. I'm just not sure how useful that would be.

Would something like what @TimZaman has proposed here be useful for you? https://github.com/NVIDIA/DIGITS/issues/667#issuecomment-220713987

**Comment by tonytan4ever** [comment]
We are currently looking at Caffe model.  I think the starting point would be modifying:

https://github.com/NVIDIA/DIGITS/blob/master/digits/model/tasks/caffe_train.py

To have an extra validation layer, maybe named validation2, is this on the right track ?

**Comment by gheinrich** [comment]
Yes, this is the right place to look at if you want to add a second validation set.

However I encourage you not to do that:
- the purpose of the training set is to optimize the model parameters through gradient descent
- the purpose of the validation set is to optimize hyper-parameters (number of epochs, learning rate, batch size, ...) by picking those that yield the best score (on the validation set)
- the purpose of the test set is to exercise your trained network on the real data

Using the test set during training is likely to yield to over-fitting. I know it's hard but you must resist that temptation!

If you see a large discrepancy between results on the validation set and on the test set then it means your validation set is not representative enough. The solution is to rework the validation set. You shouldn't have to rely on the test set at any point during training.
